[xbmc] Start work on getting rid of our emulated GetDiskFreeSpaceEx method

Paxxi opened the pull request:
Decided to unify some of our low level file handling
methods on the std::filesystem idea. As it's not standardized
yet and implementations are experimental this is a simple
version that suits our current needs. The idea is that once platforms
start offering non-experimental versions of std::filesytem we can
easily move over to using them.

<!--- Provide a general summary of your change in the Title above -->

## Description
<!--- Describe your change in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->
runtime tested on windows

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Paxxi committed:
Get rid of our emulated GetDiskFreeSpaceEx method

Decided to unify some of our low level file handling
methods on the std::filesystem idea. As it's not standardized
yet and implementations are experimental this is a simple
version that suits our current needs. The idea is that once platforms
start offering non-experimental versions of std::filesytem we can
easily move over to using them.

Paxxi commented:
jenkins build this please

MartijnKaijser commented:
jenkins build this please

Paxxi commented:
@Montellese thoughts?
maybe @wsnipex for linux and @koying for android and @Memphiz for osx

wsnipex commented on xbmc/platform/posix/Filesystem.cpp:
```diff
@@ -0,0 +1,75 @@
+  auto result = statfs(CSpecialProtocol::TranslatePath(path).c_str(), &fsInfo);
+#else
+  struct statfs64 fsInfo;
+  auto result = statfs64(CSpecialProtocol::TranslatePath(path).c_str(), &fsInfo);
```
quote from the linux standards base: https://refspecs.linuxbase.org/LSB_5.0.0/LSB-Core-generic/LSB-Core-generic/baselib-statfs64.html
>Note: Application developers should use the statvfs64() function to obtain general file system information. Applications should only use the statfs64() function if they must determine the file system type, which need not be provided by statvfs64().

Paxxi replied:
I just copied the code from current implementation, will fix

koying commented on xbmc/utils/SystemInfo.cpp:
```diff
@@ -451,67 +453,64 @@ const std::string& CSysInfo::GetAppName(void)
-      ULTotalTmp.QuadPart+= ULTotal.QuadPart;
-      ULTotalFreeTmp.QuadPart+= ULTotalFree.QuadPart;
-    }
+    total = space("/", ec);
```
This actually makes no sense on unix-derived platforms ;)
Not your fault, obviously

Paxxi replied:
Yeah it should probably fetch the mounts and iterate over them as on windows?

koying replied:
Yeah, but even that doesn't make much sense to get "Free Disk Space"...
What is Windows doing when not passed an actual path?

Paxxi replied:
Windows only cares about a disk or mount so we pass it c: d: and so on. 
Possible to mount in directories like Linux but afaik I've never seen it used on windows so we don't support it.  

Suggestions on what would make sense?

Paxxi replied:
Right, looking at IStorageProvider the GetDiskUsage method might be suitable to use but maybe add an overloaded version that returns the numbers and not formatted strings

koying replied:
> Suggestions on what would make sense

No, really...
Seems only used from https://github.com/xbmc/xbmc/blob/61f212f75c69a7478b93cf87075dc305734320b4/xbmc/utils/SystemInfo.cpp#L1045 and the whole concept of "GetHddSpaceInfo" seems bogus to me ;)

koying replied:
The thing is untouched since years. I bet it still dates from single disk / single partition xbox ;)

Paxxi replied:
and that method is called from CGUIInfoManager so it's likely that some random skin out there uses it, I'm all for dropping it but maybe @phil65 or @ronie has something to say about it.

ronie replied:
correct, there are a number of skins that use the System.FreeSpace / System.TotalSpace infolabels.

they are either shown in some home screen widget or in the filemanager window.

koying commented:
Wouldn't hooking into our StorageProvider a valid alternative?

Paxxi commented:
Our storage provider is a bit higher in the hierarchy for what I want to achieve with the filesystem methods, intending to implement stat and possibly the directory iterators so we have a common low level abstraction and can reduce some ifdefs. 

Using storage manager for the sysinfo might be a good idea, will have a look at it and see.

thebrid commented on xbmc/platform/posix/Filesystem.cpp:
```diff
@@ -0,0 +1,75 @@
+  if (result != 0)
+  {
+    ec.assign(result, std::system_category());
+    sp.available = static_cast<uintmax_t>(-1);
```
Consider using std::numeric_limits<uintmax_t>::max()

Paxxi replied:
standards draft uses static_cast http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0218r0.html#space_info

Doesn't make any real difference but avoids an include :)

thebrid replied:
:+1:

Paxxi committed:
Use statvfs64 on linux and some rearrangements

wsnipex commented on xbmc/platform/posix/Filesystem.cpp:
```diff
@@ -0,0 +1,70 @@
+    sp.free = static_cast<uintmax_t>(-1);
+    return sp;
+  }
+  sp.available = static_cast<uintmax_t>(fsInfo.f_bavail * fsInfo.f_bsize);
```
note: while it's probably negligible for our purpose, but this is incorrect when kodi is run as root, because  file systems reserve a certain percentage of the disk space for root(usually 5%-10%)
I'd still keep it as is, because distros should really stop running kodi as root

Paxxi replied:
Reading https://www.ibm.com/support/knowledgecenter/ssw_i5_54/apis/statvf64.htm it seems like a strange design that it specifically mentions non-privileged users. I'm thinking it should report the correct amount depending on user and quotas but maybe that's left to another api?

The draft standards mentions it should report the f_bavail * f_bsize so we should be good in that regard http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0218r0.html#space_info

stefansaraev replied:
first of all, not all filesystem reserve root-only blocks. and second, you dont have to care about that. no matter if kodi is run as root or not. those blocks are not to be reported in a different way depending on current uid. even fs specific tools like "df" dont care, you should not. too.

Paxxi replied:
Thank you for the explanation, at least we're all in agreement on this line of code. Let the battle over root vs nonroot continue in all its glory 😊

Memphiz commented on xbmc/platform/posix/Filesystem.cpp:
```diff
@@ -0,0 +1,70 @@
+
+#if defined(TARGET_LINUX)
+#include <sys/statvfs.h>
+#elif defined(TARGET_DARWIN) && defined(TARGET_FREEBSD)
```
Should this be an "or" instead?

Paxxi replied:
It certainly should!

Memphiz commented:
Looks straight forward to me

Paxxi committed:
Drop the idea of total space from GetDiskSpace.

It's wrong on all systems in different ways so let's
keep it fairly simple and return only for a specified mount
or default to / or C:

Paxxi commented:
Fixed the invalid test for TARGET_BSD / TARGET_DARWIN

Dropped the counting of "total" space on win32 as it's a useless metric, added some comments but kept the GetDiskspaceInfo method for backward compat with skins that rely on it.

jenkins build this please

Paxxi commented:
jenkins build this please

Paxxi commented:
build error unrelated.
